webapi/src/main/scala/org/knora/webapi/responders/admin/PermissionsResponderADM.scala

```diff
@@ -477,43 +485,91 @@ class PermissionsResponderADM(responderData: ResponderData) extends Responder(re
     }
 
     /**
+     * Adds a new administrative permission (internal use).
      *
-     * @param newAdministrativePermissionV1
-     * @param userProfileV1
-     * @return
+     * @param createRequest     the administrative permission to add.
+     * @param requestingUser    the requesting user.
+     * @param apiRequestID      the API request ID.
+     * @return an optional [[AdministrativePermissionADM]]
      */
-    private def administrativePermissionCreateRequestADM(newAdministrativePermissionV1: NewAdministrativePermissionADM, userProfileV1: UserADM): Future[AdministrativePermissionCreateResponseADM] = {
-        //log.debug("administrativePermissionCreateRequestADM")
-
-        // FIXME: Check user's permission for operation (issue #370)
-
-        for {
-            _ <- Future {
-                // check if necessary field are not empty.
-                if (newAdministrativePermissionV1.forProject.isEmpty) throw BadRequestException("Project cannot be empty")
-                if (newAdministrativePermissionV1.forGroup.isEmpty) throw BadRequestException("Group cannot be empty")
-                if (newAdministrativePermissionV1.hasNewPermissions.isEmpty) throw BadRequestException("New permissions cannot be empty")
-            }
-
-            checkResult <- administrativePermissionForProjectGroupGetADM(newAdministrativePermissionV1.forProject, newAdministrativePermissionV1.forGroup, requestingUser = KnoraSystemInstances.Users.SystemUser)
-
-            _ = checkResult match {
-                case Some(ap) => throw DuplicateValueException(s"Permission for project: '${newAdministrativePermissionV1.forProject}' and group: '${newAdministrativePermissionV1.forGroup}' combination already exists.")
-                case None => ()
-            }
+    private def administrativePermissionCreateADM(createRequest: CreateAdministrativePermissionAPIRequestADM,
```

Comment:
Please rename the method to `administrativePermissionCreateRequestADM `. There is actually a naming convention in the whole admin part of the API where messages used in routes have `request` in them. Messages with `request` in their name also always return a message with `response` in their name or an exception.

Messages without `request` in their name usually return an option and should not return an exception. They are only used internally and sent between responders/store.

In general, a method handling a certain message should have the same name as the message.

Reply:
> Please rename the method to `administrativePermissionCreateRequestADM `. There is actually a naming convention in the whole admin part of the API where messages used in routes have `request` in them. Messages with `request` in their name also always return a message with `response` in their name or an exception.
> 
> Messages without `request` in their name usually return an option and should not return an exception. They are only used internally and sent between responders/store.
> 
> In general, a method handling a certain message should have the same name as the message.

Well, actually having a look at the code this convention and design is not applied consistently on the entire admin API. For example for the creation of a group, `GroupCreateRequestADM` is routed to the responder `createGroupADM` that returns a `Future[GroupOperationResponseADM]`. No internal method is used here and in fact, this method returns an exception if the group is not created.

on the other hand, there is no such an internal method defined for the creation of projects either, i.e. the messages `ProjectCreateRequestADM` is routed to the responder `projectCreateRequestADM` which directly returns `Future[ProjectOperationResponseADM]`. Even though the naming convention you mention is somehow respected here, there is no internal method without `request` in its name defined and the `projectCreateRequestADM` directly returns an exception if no project is created. 

The same inconsistency can be seen in `createNewUserADM`, which does not have the internal method and throws an exception directly.

Therefore, since I found the code easier to read without the internal method, I left it out. But I can rename the `administrativePermissionCreateADM ` to  `administrativePermissionCreateRequestADM `. Indeed the other services of admin API (user, group, project, etc.) must be refactored accordingly.

Reply:
In general, a `request` should be answered by a `response` message. Internal (without `response`) methods and messages where only added in those cases where needed. If you agree with these conventions, then we should write them somewhere down, enforce them in future reviews, and open issues for those cases where it is not like this. You are the first person that workes on the admin API besides me, so I'm very glad when you point out inconsistencies.